pkg/gadgets/snisnoop/gadget.go

```diff
@@ -0,0 +1,267 @@
+// limitations under the License.
+
+package snisnoop
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	gadgetv1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/apis/gadget/v1alpha1"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets"
+	snitracer "github.com/kinvolk/inspektor-gadget/pkg/gadgets/snisnoop/tracer"
+	snitypes "github.com/kinvolk/inspektor-gadget/pkg/gadgets/snisnoop/types"
+	pb "github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/api"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/containerutils"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/pubsub"
+	eventtypes "github.com/kinvolk/inspektor-gadget/pkg/types"
+)
+
+type Trace struct {
+	resolver gadgets.Resolver
+	client   client.Client
+
+	started bool
+
+	tracer *snitracer.Tracer
+
+	netnsHost uint64
+}
+
+type TraceFactory struct {
+	gadgets.BaseFactory
+
+	netnsHost uint64
+}
+
+func NewFactory() gadgets.TraceFactory {
+	netnsHost, _ := containerutils.GetNetNs(os.Getpid())
+	return &TraceFactory{
+		BaseFactory: gadgets.BaseFactory{DeleteTrace: deleteTrace},
+		netnsHost:   netnsHost,
+	}
+}
+
+func (f *TraceFactory) Description() string {
+	return `The snisnoop gadget retrieves Server Name Indication (SNI) from TLS requests.`
+}
+
+func (f *TraceFactory) OutputModesSupported() map[string]struct{} {
+	return map[string]struct{}{
+		"Stream": {},
+	}
+}
+
+func deleteTrace(name string, t interface{}) {
+	trace := t.(*Trace)
+	if trace.started {
+		trace.resolver.Unsubscribe(genPubSubKey(name))
+		trace.tracer.Close()
+		trace.tracer = nil
+	}
+}
+
+func (f *TraceFactory) Operations() map[string]gadgets.TraceOperation {
+	n := func() interface{} {
+		return &Trace{
+			client:    f.Client,
+			resolver:  f.Resolver,
+			netnsHost: f.netnsHost,
+		}
+	}
+
+	return map[string]gadgets.TraceOperation{
+		"start": {
+			Doc: "Start snisnoop",
+			Operation: func(name string, trace *gadgetv1alpha1.Trace) {
+				f.LookupOrCreate(name, n).(*Trace).Start(trace)
+			},
+		},
+		"stop": {
+			Doc: "Stop snisnoop and store results",
+			Operation: func(name string, trace *gadgetv1alpha1.Trace) {
+				f.LookupOrCreate(name, n).(*Trace).Stop(trace)
+			},
+		},
+	}
+}
+
+type pubSubKey string
+
+func genPubSubKey(name string) pubSubKey {
+	return pubSubKey(fmt.Sprintf("gadget/snisnoop/%s", name))
+}
+
+func (t *Trace) Start(trace *gadgetv1alpha1.Trace) {
+	if t.started {
+		trace.Status.OperationError = ""
+		trace.Status.Output = ""
+		trace.Status.State = "Started"
+		return
+	}
+
+	var err error
+	t.tracer, err = snitracer.NewTracer()
+	if err != nil {
+		trace.Status.OperationError = fmt.Sprintf("Failed to start sni tracer: %s", err)
+		return
+	}
+
+	fillEvent := func(event *snitypes.Event, key string) {
+		keyParts := strings.SplitN(key, "/", 2)
+		if len(keyParts) == 2 {
+			event.Namespace = keyParts[0]
+			event.Pod = keyParts[1]
+		} else if key != "" {
```

Comment:
I was getting a lot of `Error on node ubuntu-hirsute: unknown key host` messages. I changed this to compare against `"host"` and now it works fine. 

Btw, I think we have the same problem in the dns gadget: https://github.com/kinvolk/inspektor-gadget/blob/370871504c53dc81e2ed25d4e4f57ba303a720cf/pkg/gadgets/dns/gadget.go#L131